From 424ed93c06c578921ac6d957da2cbdf9ebf906b4 Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Sat, 27 Sep 2014 15:42:50 +0100
Subject: [PATCH 11/14] MSYS2: Changes to various system calls

openRootTerminal(), runCommandInTerminal()
getSUCommand(), isRootRunning()

MSYS2 behaviour changes for these:
*Terminal() runs msys2_shell.cmd
getSUCommand() returns "sh.exe -c "
isRootRunning() returns false
---
 src/constants.h   |  1 +
 src/terminal.cpp  | 12 +++++++++++-
 src/unixcommand.h |  6 ++++++
 src/wmhelper.cpp  |  5 ++++-
 4 files changed, 22 insertions(+), 2 deletions(-)

diff --git a/src/constants.h b/src/constants.h
index 59ab7e8..3e0898f 100644
--- a/src/constants.h
+++ b/src/constants.h
@@ -99,6 +99,7 @@ const int ctn_RUN_IN_TERMINAL(328);
 //WMHelper related
 const QString ctn_NO_SU_COMMAND("none");
 const QString ctn_ROOT_SH("/bin/sh -c ");
+const QString ctn_MSYS2_SH("cmd.exe /c ");
 
 const QString ctn_LXQTSU("lxqt-sudo");
 
diff --git a/src/terminal.cpp b/src/terminal.cpp
index 1277809..028aaae 100644
--- a/src/terminal.cpp
+++ b/src/terminal.cpp
@@ -205,7 +205,12 @@ void Terminal::openRootTerminal()
 {
   if (m_selectedTerminal == ctn_AUTOMATIC)
   {
-    if (UnixCommand::getLinuxDistro() == ectn_MOOOSLINUX && UnixCommand::hasTheExecutable(ctn_RXVT_TERMINAL))
+    if (UnixCommand::getLinuxDistro() == ectn_MSYS2)
+    {
+        QString cmd = ctn_MSYS2_SH + UnixCommand::getMSYS2Root() + QLatin1String("/msys2_shell.cmd");
+        m_process->startDetached(cmd);
+    }
+    else if (UnixCommand::getLinuxDistro() == ectn_MOOOSLINUX && UnixCommand::hasTheExecutable(ctn_RXVT_TERMINAL))
     {
       QString cmd = WMHelper::getSUCommand() + " \"" + ctn_RXVT_TERMINAL +
           " -name Urxvt -title Urxvt \"";
@@ -359,6 +364,11 @@ void Terminal::runCommandInTerminal(const QStringList &commandList)
 
   if (m_selectedTerminal == ctn_AUTOMATIC)
   {
+    if (UnixCommand::getLinuxDistro() == ectn_MSYS2)
+    {
+      QString cmd = ctn_MSYS2_SH + UnixCommand::getMSYS2Root() + QLatin1String("/msys2_shell.cmd -c ") + ftemp->fileName();
+      m_process->start(cmd);
+    }
     if (UnixCommand::getLinuxDistro() == ectn_MOOOSLINUX && UnixCommand::hasTheExecutable(ctn_RXVT_TERMINAL))
     {
       QString cmd =
diff --git a/src/unixcommand.h b/src/unixcommand.h
index 3d09880..3fadb16 100644
--- a/src/unixcommand.h
+++ b/src/unixcommand.h
@@ -113,10 +113,16 @@ public:
   static bool hasTheExecutable( const QString& exeName );
   static bool isAppRunning(const QString &appName, bool justOneInstance = false);
 
+#if defined(_WIN32)
+  static bool isRootRunning(){
+    return false;
+  }
+#else
   static bool isRootRunning(){
     int uid = geteuid();
     return (uid == 0); //Returns TRUE if root is running Octopi
   }
+#endif
 
   static QString getMSYS2Root(){
     if (m_msys2Root.isEmpty())
diff --git a/src/wmhelper.cpp b/src/wmhelper.cpp
index a373b8b..aa4d7e0 100644
--- a/src/wmhelper.cpp
+++ b/src/wmhelper.cpp
@@ -309,7 +309,10 @@ QString WMHelper::getLXQTSUCommand(){
 QString WMHelper::getSUCommand(){
   QString result(ctn_NO_SU_COMMAND);
 
-  if (isXFCERunning() && (UnixCommand::hasTheExecutable(ctn_GKSU_2))){
+  if (UnixCommand::getLinuxDistro() == ectn_MSYS2){
+    return QLatin1String("sh.exe -c ");
+  }
+  else if (isXFCERunning() && (UnixCommand::hasTheExecutable(ctn_GKSU_2))){
     result = getGKSUCommand();
   }
   else if (isKDERunning() && UnixCommand::hasTheExecutable(ctn_KDESU)){
-- 
2.8.2

